base/lock.jl

```diff
@@ -556,3 +556,59 @@ end
     import .Base: Event
     export Event
 end
+
+const _NOT_CALLED = Ptr{Nothing}(0)
+const _STARTED = Ptr{Nothing}(1)
+const _DONE = Ptr{Nothing}(2)
+
+"""
+    Once()
+
+Create a `Once` object that can be used to execute a callback exactly once across all executions of the program.
+The callback is executed by calling `call_once!`. This is similar to `pthread_once` in POSIX.
+
+```
+    const once = Once()
+    foo() = println("This will be printed exactly once")
+    call_once!(once, foo)
+    call_once!(once, foo) # does nothing
+```
+
+!!! compat "Julia 1.12"
+    This functionality requires at least Julia 1.12
+"""
+mutable struct Once
+    @atomic done::Ptr{Nothing} # Ptr is used so it resets during precompilation
+    lock::Threads.Event #Ideally would be a Futex
```

Comment:
Seems like this should be a Condition variable, since being an Event is redundant with `@done`
```suggestion
    lock::Threads.Condition
```

Reply:
Does Condition actually do what we want here? Event is here serving just as something for threads to sleep in if they enter the slow path. Specifically I want something that if somehow enters the slow path to go right through it.

Reply:
You need to call it with `@lock cond while !once.done; wait(cond); end`

Reply:
Right. But does it allow anyone to go through once it's being notified? Because that's the behaviour I want that Event has.